woocommerce/woocommerce
# Fix the visibility of partially out of stock variable products when using the layered nav widget

### All Submissions:


### Changes proposed in this Pull Request:

This PR fixes an issue that appears when using the layered nav widget (which allows filtering the listed products by an attribute) and there are variable products where some variations have stock and others don't. The issue is that when an attribute is selected for filtering, all variable products having a variation that corresponds to that attribute will be displayed, even those not having stock for that variation (as long as at least one variation has stock).

Example: you sell shoes in two colors, black and blue. Black is out of stock but there's stock for blue. You select "black" (but not "blue") in the widget. The shoes will be included in the products list, but they shouldn't.

Additionally, the count of products that appear in the widget for each attribute is also wrongly including the variable products that don't have stock for the attribute corresponding to that variation.

Note that the original issue is about products out of stock, but the changes act actually on the more wide visibility condition. A product is "visible" if it's published, not excluded for catalog, and has stock. Additionally, a variable product will not be considered visible if the parent product is not.

#### New behavior of the counters

Now the counters count all the variations that are in stock, regardless of whether the filter uses OR or AND logic; that's because one given variation has always one single value for the attribute, so the logic in use is irrelevant. For non-variable products they continue working as usual.

Variation counts will be equivalent to product counts when one single attribute is used to define variations, but if more than one attribute is used that could not be the case. For example you have a "shoes" product with two attributes used for variations, "color" and "style"; if you define variations as "black classic" and "black sport" then "black" will be counted twice in the filter (if both variations have stock).

Also, when using multiple filters they interact with AND logic as usual: only variations that have both attributes in its definition are counted.

Closes #25524.

### How to test the changes in this Pull Request:

**NOTE:** You need to run database updates for the widget counters to work properly!

1. Go to Settings > Products > Inventory, mark "Hide out of stock items from the catalog".
2. Create an attribute named "Color", with the following terms: Black, Brown, Blue, Green, Pink.
3. Create the following variable products: Big shoes, Medium shoes, Small shoes, Kid shoes. Create variations for all the existing colors on all of them. Assign any regular price to all the variations. (Hint: create one product, then duplicate)
4. Modify the products so the following variations are in stock and the others are out of stock:
* Big: black, brown.
* Medium: blue, brown.
* Small: blue, green.
* Kids: green, pink.
5. Add the "Filter by attribute" widget, filtering by color with OR.
6. Try filtering by one or two colors and verify that the results make sense. For example when filtering by "blue" you should see only "Medium shoes" and "Small shoes" (without the fix you would see all four shoe types). Verify also that the results count that appear at the beginning and the end of the list is correct.
7. Change the widget filtering type to AND and try different combinations again. Verify that selecting pink and black displays the "No products available" message.

#### How to test the counters

With the same setup, verify that the counters display the following, no matter what attributes you select in the widget:

![image](https://user-images.githubusercontent.com/937723/81082993-4b04ed80-8ef4-11ea-8d0d-6e8c5a14d432.png)

Now make sure that nothing is selected, then go and change the widget operator to AND. Reload the products page and verify that the behavior is the same.

#### Testing counters with multiple attributes

1. Create an attribute named "Style" with two terms: Classic, and Sport.
2. Modify all the products so that all of them use the new style for variations, but leave the value for all variations as "Any style".
3. Modify the Big shoes product so that the Black variation is Classic, and create a new variation that is Black Sport (leave both in stock).
4. Modify the Medium shoes product so that the Brown variation is Sport (and still has stock).
5. Add the "Filter by attribute" widget, filtering by style with any logic.
6. Reload the page and verify that the counters now appear like this:

![image](https://user-images.githubusercontent.com/937723/87396534-42dfc300-c5b3-11ea-8bb0-7f040e5c2206.png)

Note how there are two variations listed for Black, since we added a new one to the Big shoes. Also note how the two classic variations and the single sport variation are listed in the style filter.

7. Select "Black" in the color filter and note how the style filter now displays a value of 1 for classic and 1 for sport, since there's one visible variation of each.
8. Unselect all and select "Blue" in the color filter and note how the style filter disappears, since there's no variation that is blue and has a style associated.
9. Unselect all in the color filter and select "Sport" in the style filter, you will see this:

![image](https://user-images.githubusercontent.com/937723/87396197-c1883080-c5b2-11ea-9720-7e86fe4d298f.png)

#### Testing visibility with multiple filters

While testing counters with multiple filters you may have noticed that regardless of what filters you selected in the style filter you were still seeing all the products in the catalog (if you don't filter by color). That's because all of the products have visible variations with "Any style" and thus they always match the style filter.

Modify the "Kid shoes" product so that all the variations have the sport style, then select "Classic" in the style filter while not filtering by color, and you'll see that the product disappears from the catalog.

### Other information:


### Changelog entry

> Fix - The filtering widget now works as expected with variable products, displaying those products for which visible variations are available.
